File: 238. Product of Array Except Self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        
        n = len(nums)
        
        # key: (i, j) starting (inclusive) & ending (inclusive) idx of nums
        # val: product
        t = {
            (0, 0): nums[0],
            (n - 1, n - 1): nums[-1]
            }
    
        # calculate products of subarrays
        for j in range(1, n - 1):
            t[(0, j)] = t[(0, j - 1)] * nums[j]
        
        for i in range(n - 2, 0, -1):
            t[(i, n - 1)] = t[(i + 1, n - 1)] * nums[i]
            
        # return arr to be filled
        arr = [0 for i in range(n)]
        
        # construct the return arr
        for i in range(n):
            # multiply left subarray and right subarray
            if i == 0:
                left = 1
            else:
                left = t[(0, i - 1)]
            if i == n - 1:
                right = 1
            else:
                right = t[(i + 1, n - 1)]
            
            arr[i] = left * right
         
        return arr
            

File: 238. Product of Array Except Self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        
        // length of nums
        int n = nums.length;
        
        // create helper arr L
        int[] L = new int[n];
        // create helper arr R
        int[] R = new int[n];
        // create an array containing the products to be returned
        int[] res = new int[n];
        
        
        // fill arr L
        L[0] = 1;
        for (int i = 1; i < n; i++) {
            L[i] = L[i - 1] * nums[i - 1];
        }        
        
        // fill arr R
        R[n - 1] = 1;
        for (int j = n - 2; j >= 0; j--) {
            R[j] = R[j + 1] * nums[j + 1];
        }
        
        // fill the return array by multiplying left and right side of each element
        for (int k = 0; k < n; k++) {
            res[k] = L[k] * R[k];
        }
        
        
        
        // return arr
        return res;
    }
}